DAO/src/lib.rs

```diff
@@ -0,0 +1,788 @@
+            did_pass: false,
+            cancelled: false,
+            aborted: false,
+            token_tribute,
+            details,
+            starting_period,
+            max_total_shares_at_yes_vote: 0,
+            votes_by_member: BTreeMap::new(),
+        };
+        self.proposals.insert(self.proposal_id, proposal);
+        let new_proposal = MembershipProposal {
+            proposer: H256::from_slice(msg::source().as_ref()),
+            applicant: H256::from_slice(applicant.as_ref()),
+            proposal_id: self.proposal_id,
+            token_tribute,
+        };
+        msg::reply(
+            Event::SubmitMembershipProposal(new_proposal),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+        self.proposal_id = self.proposal_id.saturating_add(1);
+    }
+
+    // The proposal of funding
+    // Requirements:
+    // * The proposal can be submitted only by the existing members or their delegate addresses
+    // * The receiver ID can't be the zero
+    // * The DAO must have enough funds to finance the proposal
+    // Arguments:
+    // * `receiver`: an actor that will be funded
+    // * `amount`: the number of ERC20 tokens that will be sent to the receiver
+    // * `quorum`: a certain threshold of YES votes in order for the proposal to pass
+    // * `details`: the proposal description
+    async fn submit_funding_proposal(
+        &mut self,
+        applicant: &ActorId,
+        amount: u128,
+        quorum: u128,
+        details: String,
+    ) {
+        // check that `msg::source()` is either a DAO member or a delegate key
+        match self.member_by_delegate_key.get(&msg::source()) {
+            Some(member) => {
+                if !self.is_member(member) {
+                    panic!("account is not a DAO member");
+                }
+            }
+            None => {
+                panic!("account is not a delegate");
+            }
+        }
+
+        if applicant == &ZERO_ID {
+            panic!("Proposal for the zero address");
+        }
+
+        // check that DAO has sufficient funds
+        let balance = balance(&self.approved_token_program_id, &exec::program_id()).await;
+        if balance < amount {
+            panic!("Not enough funds in DAO");
+        }
+
+        let mut starting_period = exec::block_timestamp();
+        // compute startingPeriod for proposal
+        // there should be a minimum time interval between proposals (period_duration) so that members have time to ragequit
+        if self.proposal_id > 0 {
+            let previous_starting_period = self
+                .proposals
+                .get(&(&self.proposal_id - 1))
+                .unwrap()
+                .starting_period;
+            if starting_period < previous_starting_period + self.period_duration {
+                starting_period = previous_starting_period + self.period_duration;
+            }
+        }
+
+        let proposal = Proposal {
```

Comment:
I'd implement/derive `Default` trait for `Proposal` struct and use `..Default::default()` notation for default values (`0`, `false`, `BTreeMap::new()`).